### DISCUSSION TITLE
Sources and permissions portability

### DISCUSSION THREAD

**Discussion by jmauzyk** [verbb/formie#1696] [Q&A]
I've noticed a couple odd behaviors due to a form having been exported from a development environment and imported into a production environment, therefore having different IDs / UIDs in each.

In my development environment, I have the last few sources of the Submissions index arranged like this:
<img width="220" alt="Screenshot 2024-01-23 at 4 47 45 PM" src="https://github.com/verbb/formie/assets/21298596/44c808ff-b3b9-4c89-8317-09b0dccc4b6a">

However, after those settings are deployed to our production environment with `allowAdminChanges` disabled, the sources look like this due to the fact that they do not have identical IDs:
<img width="221" alt="Screenshot 2023-12-12 at 10 40 54 AM" src="https://github.com/verbb/formie/assets/21298596/40f917e3-cd28-41a0-9628-410284de2f9e">

Similarly, if I set a permission in the dev environment allowing a user group to manage a certain type of submission, that permission will not be enabled when deployed to production if the form doesn't have the same UID as in development.

Are there certain steps I need to take to ensure those settings match up in different environments with different identifiers? Is it possible to change how those settings are controlled to allow for more portability between different environments which may have differing identifiers?

**Comment by engram-design** [comment]
Should be fixed for the next release. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.1.2"`. Note that you'll need to re-export your forms, as we didn't previously include the UID in export data.

**Reply by jmauzyk** [reply]
Great, thanks so much!

**Reply by jmauzyk** [reply]
This fixed the issue with settings, but the out of order sources in the Submissions index is still an issue. Looks like the keys are based on IDs rather than UIDs. Is it possible to fix this as well?

**Reply by engram-design** [reply]
@jmauzyk I'm not a massive fan of changing the order to use the UID. Currently, when you add a new form, it'll show up at the bottom, which makes logical sense. If we switch to ordering by UID, they'll appear at random (being sorted by their UID).

Besides, this is also the _default_ order of the element index, as you can customise that. This order is also supposed to be stored in project config. So any changes you make to the order of items should be mirrored on other environments.

**Reply by jmauzyk** [reply]
That’s the problem I’m running into, though. When I customize the order of sources for the Submissions index in our dev environment, no matter where I put the Tax Exemption Request in the order, when the project config is applied to our production environment the Tax Exemption Request is shown separately at the end, as seen in the second image. Because the form was imported into the production environment, it has a different ID that doesn’t match the one in the project config, so it gets thrown out of order.

**Reply by engram-design** [reply]
Right, gotcha. I think we might need to come up with something else. Making the change to UIDs for the source key would alter **everyone's** element indexes, which isn't great.

**Reply by jmauzyk** [reply]
I checked the behavior of other Craft indexes to see their behavior when UID is used as the key. When no order is set for the element sources, the sources seem to automatically be shown in the order they were created. Similarly, when there is an order set of the element sources, new sources are shown in a separate section at the bottom in the order they were created. So I don't think using UIDs will cause issues with the desired behavior you mentioned.

Regarding modifying everyone's element indexes, would it be possible to avoid this by running a migration to replace the IDs with UIDs in their project config?

I'm happy to work on this and submit a PR if you're open to it.

**Reply by engram-design** [reply]
Happy for a PR, but still thinking this needs to be locked to Formie 3, as I can't see it _not_ being a change in behaviour from the current setup. The moment we switch IDs for UIDs it's going to re-order things.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.7](https://github.com/verbb/formie/releases/tag/4.0.0-beta.7). User permissions have been given an overhaul with the addition of Form Groups.
